sdk/schemaregistry/azure-data-schemaregistry-jsonschema/src/main/java/com/azure/data/schemaregistry/jsonschema/SchemaRegistrySchemaCache.java

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.data.schemaregistry.jsonschema;
+
+import com.azure.core.util.CoreUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.logging.LogLevel;
+import com.azure.data.schemaregistry.SchemaRegistryAsyncClient;
+import com.azure.data.schemaregistry.models.SchemaFormat;
+import com.azure.data.schemaregistry.models.SchemaProperties;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static com.azure.core.util.FluxUtil.monoError;
+
+/**
+ * An LRU cache of schemas.  If the schema is not yet fetched, performs the network operation and caches it.
+ */
+class SchemaRegistrySchemaCache {
+    private static final String SIZE_KEY = "size";
+    private static final String TOTAL_LENGTH_KEY = "totalLength";
+
+    private final ClientLogger logger = new ClientLogger(SchemaRegistrySchemaCache.class);
+
+    private final SchemaCache cache;
+    private final SchemaRegistryAsyncClient schemaRegistryClient;
+    private final String schemaGroup;
+    private final boolean autoRegisterSchemas;
+    private final Object lock = new Object();
+
+    SchemaRegistrySchemaCache(SchemaRegistryAsyncClient schemaRegistryClient, String schemaGroup,
+        boolean autoRegisterSchemas, int capacity) {
+
+        this.schemaRegistryClient = schemaRegistryClient;
+        this.schemaGroup = schemaGroup;
+        this.autoRegisterSchemas = autoRegisterSchemas;
+        this.cache = new SchemaCache(capacity);
+    }
+
+    Mono<String> getSchemaId(String schemaFullName, String schemaDefinition) {
+        final String existingSchemaId;
+        synchronized (lock) {
+            existingSchemaId = cache.getSchemaId(schemaDefinition);
+        }
+
+        if (existingSchemaId != null) {
+            return Mono.just(existingSchemaId);
+        }
+
+        // It is possible to create the serializer without setting the schema group. This is the case when
+        // autoRegisterSchemas is false. (ie. You are only using it to deserialize messages.)
+        if (CoreUtils.isNullOrEmpty(schemaGroup)) {
+            return monoError(logger, new IllegalStateException("Cannot serialize when 'schemaGroup' is not set. Please"
+                + " set in SchemaRegistryJsonSchemaSerializer.schemaGroup when creating serializer."));
+        }
+
+        final Mono<SchemaProperties> serviceCall;
+        if (autoRegisterSchemas) {
+            serviceCall = this.schemaRegistryClient
+                .registerSchema(schemaGroup, schemaFullName, schemaDefinition, SchemaFormat.JSON);
```

Comment:
Does `registerSchema` update the schema definition if it already exists?

Reply:
No. AFAIK no other language does this either.